Remove unnecessary tensor copies

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33905 Refactor jit::Operator
* #33904 Fix either::map naming
* #33313 Improve boxed dispatch performance
* #32521 Use codegen'ed unboxing wrappers
* **#33732 Remove unnecessary tensor copies**

move and forward instead of copy

Benchmarks:
A microbenchmark calling the add operation on two tensors in a tight loop shows a 5% improvement in performance.
No visible change for a model like resnet that does more work in its kernels.

Differential Revision: [D20082642](https://our.internmc.facebook.com/intern/diff/D20082642/)

smessmer committed:
Remove unnecessary tensor copies

move and forward instead of copy

Benchmarks:
A microbenchmark calling the add operation on two tensors in a tight loop shows a 5% improvement in performance.
No visible change for a model like resnet that does more work in its kernels.

Differential Revision: [D20082642](https://our.internmc.facebook.com/intern/diff/D20082642/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove unnecessary tensor copies"


move and forward instead of copy

Benchmarks:
A microbenchmark calling the add operation on two tensors in a tight loop shows a 5% improvement in performance.
No visible change for a model like resnet that does more work in its kernels.

Differential Revision: [D20082642](https://our.internmc.facebook.com/intern/diff/D20082642/)

[ghstack-poisoned]

bhosmer reviewed (approved):
Awesome! 

And this PR feels like a pretty healthy argument for some kind of CI-time check for these sorts of regressions, if it can be done feasibly... e.g. there's zero chance I would have spotted all of these in a straight code review.

smessmer committed:
Update on "Remove unnecessary tensor copies"


move and forward instead of copy

Benchmarks:
A microbenchmark calling the add operation on two tensors in a tight loop shows a 5% improvement in performance.
No visible change for a model like resnet that does more work in its kernels.

Differential Revision: [D20082642](https://our.internmc.facebook.com/intern/diff/D20082642/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove unnecessary tensor copies"


move and forward instead of copy

Benchmarks:
A microbenchmark calling the add operation on two tensors in a tight loop shows a 5% improvement in performance.
No visible change for a model like resnet that does more work in its kernels.

Differential Revision: [D20082642](https://our.internmc.facebook.com/intern/diff/D20082642/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove unnecessary tensor copies"


move and forward instead of copy

Benchmarks:
A microbenchmark calling the add operation on two tensors in a tight loop shows a 5% improvement in performance.
No visible change for a model like resnet that does more work in its kernels.

Differential Revision: [D20082642](https://our.internmc.facebook.com/intern/diff/D20082642/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove unnecessary tensor copies"


move and forward instead of copy

Benchmarks:
A microbenchmark calling the add operation on two tensors in a tight loop shows a 5% improvement in performance.
No visible change for a model like resnet that does more work in its kernels.

Differential Revision: [D20082642](https://our.internmc.facebook.com/intern/diff/D20082642/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove unnecessary tensor copies"


move and forward instead of copy

Benchmarks:
A microbenchmark calling the add operation on two tensors in a tight loop shows a 5% improvement in performance.
No visible change for a model like resnet that does more work in its kernels.

Differential Revision: [D20082642](https://our.internmc.facebook.com/intern/diff/D20082642/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove unnecessary tensor copies"


move and forward instead of copy

Benchmarks:
A microbenchmark calling the add operation on two tensors in a tight loop shows a 5% improvement in performance.
No visible change for a model like resnet that does more work in its kernels.

Differential Revision: [D20082642](https://our.internmc.facebook.com/intern/diff/D20082642/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove unnecessary tensor copies"


move and forward instead of copy

Benchmarks:
A microbenchmark calling the add operation on two tensors in a tight loop shows a 5% improvement in performance.
No visible change for a model like resnet that does more work in its kernels.

Differential Revision: [D20082642](https://our.internmc.facebook.com/intern/diff/D20082642/)

[ghstack-poisoned]
